lib/PerformanceUtils.ts

```diff
@@ -0,0 +1,66 @@
+import lodashTransform from 'lodash/transform';
+import {deepEqual} from 'fast-equals';
+
+type UnknownObject = Record<string, unknown>;
+
+type LogParams = {
+    keyThatChanged?: string;
+    difference?: unknown;
+    previousValue?: unknown;
+    newValue?: unknown;
+};
+
+type Mapping = Record<string, unknown> & {
+    key: string;
+    displayName: string;
+};
+
+let debugSetState = false;
+
+function setShouldDebugSetState(debug: boolean) {
+    debugSetState = debug;
+}
+
+/**
+ * Deep diff between two objects. Useful for figuring out what changed about an object from one render to the next so
+ * that state and props updates can be optimized.
+ */
+function diffObject<TObject extends UnknownObject, TBase extends UnknownObject>(object: TObject, base: TBase): UnknownObject {
+    return lodashTransform(object, (result: UnknownObject, value, key) => {
+        if (deepEqual(value, base[key])) {
+            return;
+        }
+
+        if (typeof value === 'object' && typeof base[key] === 'object') {
```

Comment:
I was worried that `typeof value === 'object'` might work differently here, but I think it's fine 👍